### PR TITLE
sla_metric_by_ticket

### PR DESCRIPTION
**Pull request by muethinr**

### CODE REVIEW

File: zendesk_ticket_metrics/sla_metric_by_ticket [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ ) as ticket_metric_event
+ where rank=1
+), ticket_metric as (
+ select distinct
+  ticket_id,
+  instance_id,
+  time,
+  case 
+   when metric = "reply_time" and instance_id =1 then "first_reply_time"
+   when metric = "reply_time" and instance_id <>1 then "next_reply_time"
```
Comment (gareginordyan) [inline zendesk_ticket_metrics/sla_metric_by_ticket]: `>1`

File: zendesk_ticket_metrics/sla_metric_by_ticket [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+   else metric 
+  end as metric,
+  sla_policy_id
+  from ticket_metric_ranked
+), ticket_priority as (
+ select 
+  ticket_id,
+  value as priority,
+  updated,
+  coalesce(lead(updated) over (partition by field_name,ticket_id order by updated),'2999-12-31 23:59:59 UTC') as changed_at
```
Comment (gareginordyan) [inline zendesk_ticket_metrics/sla_metric_by_ticket]: remove `field_name` since it's filtered

Comment (gareginordyan) [inline zendesk_ticket_metrics/sla_metric_by_ticket]: `changed_at` to `valid_until`

File: zendesk_ticket_metrics/sla_metric_by_ticket [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  business_hours,
+  target
+ from (
+  select 
+   sla_policy_id,
+   metric,
+   priority,
+   target,
+   business_hours,
+   row_number() over (partition by sla_policy_id, metric, priority order by _fivetran_synced desc) as rank
```
Comment (gareginordyan) [inline zendesk_ticket_metrics/sla_metric_by_ticket]: do we need to order by `_fivetran_synced` or `sla_policy_updated_at`?

File: zendesk_ticket_metrics/sla_metric_by_ticket [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+   priority,
+   target,
+   business_hours,
+   row_number() over (partition by sla_policy_id, metric, priority order by _fivetran_synced desc) as rank 
+  from zendesk.sla_policy_metric_history
+ ) as ranks
+ where rank =1
+)
+  select 
+   ticket_metric.ticket_id,
```
Comment (gareginordyan) [inline zendesk_ticket_metrics/sla_metric_by_ticket]: `let's add the table names in front of the field names so it's easy to follow`

File: zendesk_ticket_metrics/sla_metric_by_ticket [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+   sla_policy_metric_history_ranked.metric,
+   business_hours,
+   instance_id,
+   time,
```
Comment (gareginordyan) [inline zendesk_ticket_metrics/sla_metric_by_ticket]: rename to `sla_applied_at`

File: zendesk_ticket_metrics/sla_response_time_targets [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  join zendesk.schedule on ticket_schedule.schedule_id = schedule.id
+  group by 1,2,3
+), bussiness_weeks_to_target as (
+  select 
+   metric_by_ticket.ticket_id,
+   metric_by_ticket.metric,
+   metric_by_ticket.instance_id,
+   metric_by_ticket.sla_policy_id,
+   metric_by_ticket.time,
+   business_minutes_in_week.schedule_id,
```
Comment (gareginordyan) [inline zendesk_ticket_metrics/sla_response_time_targets]: add table names here please

File: zendesk_ticket_metrics/sla_response_time_targets [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+   target,
+   cast(ceiling(metric_by_ticket.target / business_minutes_in_week.business_minutes_in_week) as int64) as bussiness_weeks_to_target
+  from zendesk.sla_metric_by_ticket
+  join business_minutes_in_week on metric_by_ticket.ticket_id = business_minutes_in_week.ticket_id
+   and metric_by_ticket.time = business_minutes_in_week.created_at
+), adding_start_end_times as (
+  select
+   ticket_id,
```
Comment (gareginordyan) [inline zendesk_ticket_metrics/sla_response_time_targets]: add table names here please

File: zendesk_ticket_metrics/sla_response_time_targets [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  window running_total as (partition by ticket_id, metric, instance_id order by week_number,start_Time_utc rows between unbounded preceding and current row)
+), breach_time as (
+  select
+   ticket_id,
+   metric,
+   instance_id,
+   sla_policy_Id,
+   schedule_id,
+   min(timestamp_add(start_of_week,interval end_time_utc+running_minutes minute)) as breach_time
+  from running_time
```
Comment (gareginordyan) [inline zendesk_ticket_metrics/sla_response_time_targets]: rename to `...left`

File: zendesk_ticket_metrics/sla_response_time_targets [hunk 9]
```diff
@@ -0,0 +1,85 @@
+CREATE VIEW zendesk.sla_response_time_targets as (
+with business_minutes_in_week as (
+  select
+   ticket_schedule.ticket_id,
+   ticket_schedule.created_at,
+   schedule.id as schedule_id, 
+   sum(end_time_utc - start_time_utc) as business_minutes_in_week
+  from zendesk.ticket_schedule
```
Comment (gareginordyan) [inline zendesk_ticket_metrics/sla_response_time_targets]: split into two per our discussion

**Comment by gareginordyan** [comment]
image

**Comment by gareginordyan** [comment]
why are there so many missing ones?

File: zendesk_ticket_metrics/sla_metric_by_ticket (hunk 1 above)
Comment (muethinr) [reply zendesk_ticket_metrics/sla_metric_by_ticket]: updated

File: zendesk_ticket_metrics/sla_metric_by_ticket (hunk 2 above)
Comment (muethinr) [reply zendesk_ticket_metrics/sla_metric_by_ticket]: updated

Comment (muethinr) [reply zendesk_ticket_metrics/sla_metric_by_ticket]: updated

File: zendesk_ticket_metrics/sla_metric_by_ticket (hunk 3 above)
Comment (muethinr) [reply zendesk_ticket_metrics/sla_metric_by_ticket]: changed to sla_policy_updated_at

File: zendesk_ticket_metrics/sla_metric_by_ticket (hunk 4 above)
Comment (muethinr) [reply zendesk_ticket_metrics/sla_metric_by_ticket]: updated

File: zendesk_ticket_metrics/sla_metric_by_ticket (hunk 5 above)
Comment (muethinr) [reply zendesk_ticket_metrics/sla_metric_by_ticket]: updated

**Comment by gareginordyan** [comment]
@fraserharris I don't know enough about the SLA metrics and have not done enough research on them.  I provided some general feedback and @muethinr made the suggested adjustments.  Would you or someone from your team be able to review to make sure these satisfy your requirements before merging?

**Comment by muethinr** [comment]
Here is a compare of current results: https://docs.google.com/spreadsheets/d/1ucV4W-1QwwrA1nnRxZrRJ0JlwaLOCDNgo_5OdrY90MY/edit#gid=1144959937


there are still alot tickets in the existing sla_policy_breach that have null values

**Comment by fraserharris** [comment]
Cool - this is looking pretty good. @muethinr I think we need to go through some of them manually.  I'm 90% the majority of differences are due to: 1) incorrect Time Zone and 2) there _was_ a breached_at time when the ticket was sync'ed, but subsequently an event has happened that clears the timer.

File: zendesk_ticket_metrics/sla_metric_by_ticket [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+     select 
+      instance_id,
+      metric,
+      time,
+      sla_policy_id,
+      ticket_id,
+      row_number() over (partition by instance_id,metric,ticket_id order by time asc) rank
+     from zendesk.ticket_metric_event
+     where type='apply_sla'
+   ) as ticket_metric_event
```
Comment (fraserharris) [inline zendesk_ticket_metrics/sla_metric_by_ticket]: We are eliminating this table - why is the query based on this?  Recall I asked you to create a `ticket_sla_policy` VIEW on top of `ticket_metric_event`

Comment (muethinr) [reply zendesk_ticket_metrics/sla_metric_by_ticket]: I was under the impression that `sla_policy_breach` was the one that was being removed. I'll put some time on your calendar so we can review.

**Comment by muethinr** [comment]
approved per @fraserharris

**Pull request merged**